### PR TITLE
rev=true in searchsorted_interval

### PR DESCRIPTION
**Pull request by aplavin**
this also simplifies findall

no new tests added: rely on the extensive findall testsuite

### CODE REVIEW

**Review by hyrodium — COMMENTED** [review]
I found some performance degression. Could you check the following result?
```julia
julia> versioninfo()
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD Ryzen 7 2700X Eight-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, znver1)
```

**On the current master**
```julia
julia> @benchmark findall(∈(1..3), 2:8)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.020 ns … 0.031 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.020 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.021 ns ± 0.004 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▂
  0.02 ns        Histogram: frequency by time       0.03 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**With this PR**
```julia
julia> @benchmark findall(∈(1..3), 2:8)
BenchmarkTools.Trial: 10000 samples with 870 evaluations.
 Range (min … max):  136.305 ns …  2.080 μs  ┊ GC (min … max): 0.00% … 92.90%
 Time  (median):     141.348 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   147.630 ns ± 89.578 ns  ┊ GC (mean ± σ):  2.95% ±  4.52%

   ▁▄▆████▇▆▅▃▂▁▁                               ▁        ▁ ▁   ▃
  ▇████████████████▆▆▆▅▅▅▄▅▄▅▅▄▅▄▅▅▁▄▃▅▅▅▄▁▅▆▆▇███▇▇▇████████▇ █
  136 ns        Histogram: log(frequency) by time       183 ns <

 Memory estimate: 128 bytes, allocs estimate: 4.
```

**Comment by hyrodium** [comment]
>This PR also fixes https://github.com/JuliaMath/IntervalSets.jl/issues/100 btw.

Great! Could you add the following tests in `test/findall.jl`?

```julia
    @testset "zero-step ranges" begin
        r = range(1,1,10)
        @test findall(in(1..3),r) == 1:10
        @test findall(in(2..3),r) |> isempty
    end
```

>Basically, the issue is that without lt = < Base searchsorted functions use isless, that doesn't consider +-0 to be equal:

Thanks! I didn't know that property of `searchsorted`.

>and keeps the performance for ranges (Base searchsorted funcs are called without lt=< argument).

Hmm, the performance is now getting much better than the first benchmark, but there still be some performance regressions:

**On the current master**

```julia
julia> using IntervalSets, BenchmarkTools

julia> @benchmark findall(∈(1..3), 2:8)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.020 ns … 0.031 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.020 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.021 ns ± 0.004 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                          
  █▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▂
  0.02 ns        Histogram: frequency by time       0.03 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark findall(∈($(1..3)), $(2:8))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.222 ns … 18.574 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.332 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.383 ns ±  0.229 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                   ▅▃█ ▄                                       
  ▂▂▁▂▁▂▁▂▂▂▂▁▄▁▇▁▃███▁█▁▂▁▂▂▁▁▁▁▁▁▁▁▂▁▃▁▅▂▂▆▁▅▁▃▁▂▂▂▂▁▂▁▅▁▄▃ ▃
  11.2 ns         Histogram: frequency by time        11.6 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**With this PR**

```julia
julia> using IntervalSets, BenchmarkTools

julia> @benchmark findall(∈(1..3), 2:8)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.442 ns … 5.931 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.472 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.471 ns ± 0.097 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▄  █                                                      
  ██▂▃█▃▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▂▁▂▁▁▁▁▁▂▂▁▂▂▁▂▂▁▂▂▁▁▂▁▂▂▁▂▂▂ ▂
  1.44 ns        Histogram: frequency by time       1.82 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark findall(∈($(1..3)), $(2:8))

BenchmarkTools.Trial: 10000 samples with 927 evaluations.
 Range (min … max):  112.177 ns …  1.880 μs  ┊ GC (min … max): 0.00% … 93.46%
 Time  (median):     119.698 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   125.097 ns ± 87.030 ns  ┊ GC (mean ± σ):  3.47% ±  4.66%

    ▂▄▅▅▅▇▇▇▇█▇▄▃▂▂▁                                  ▁        ▂
  ▆██████████████████▆▇▆▅▆▅▄▆▅▄▃▃▄▁▄▁▄▄▆▅▆▆▇▆▅▆▆▆▆▇███████▇▅▆▇ █
  112 ns        Histogram: log(frequency) by time       157 ns <

 Memory estimate: 128 bytes, allocs estimate: 4.
```
Sorry I have not taken the time to review. I have not yet checked the reasons for the performance degression.

**Comment by hyrodium** [comment]
I checked the benchmarks again, and it seems there still be some performance degressions with `findall`.

**Before this PR**

```julia
julia> using IntervalSets, BenchmarkTools

julia> i = 1..3
1..3

julia> r = 2:8
2:8

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.332 ns … 18.383 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.443 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.496 ns ±  0.203 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

       ▁▃    ▃ ▂  ▇ █  ▂ ▂  ▁ ▁ ▂  ▂ ▆    ▅▂▅               ▃ ▂
  ▇▆▁▅▁██▁▃▁▁█▁█▁██▁█▁▆█▁█▁▆█▁█▁█▅▇█▁█▅█▇▁████▁▆▁▃▄▁▇▁▆▇▁▅▁▇█ █
  11.3 ns      Histogram: log(frequency) by time      11.7 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> r = 2.0:8.0
2.0:1.0:8.0

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 980 evaluations.
 Range (min … max):  62.996 ns … 186.444 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     64.929 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   65.645 ns ±   5.061 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁▆▇█▆▁▁▁▁▁                       ▂                           ▂
  ███████████▇▅▅▄▄▄▄▄▃▃▁▃▁▁▃▁▁▃▁▃▆▄█▁▄▃▁▃▄▄▁▁▄▄▃▆▅▄▅▃▃▄▃▅▁▅▄▅▅ █
  63 ns         Histogram: log(frequency) by time      99.6 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**After this PR**

```julia
julia> using IntervalSets, BenchmarkTools
[ Info: Precompiling IntervalSets [8197267c-284f-5f27-9208-e0e47529a953]

julia> using IntervalSets, BenchmarkTools

julia> i = 1..3
1..3

julia> r = 2:8
2:8

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 979 evaluations.
 Range (min … max):  64.330 ns …  1.845 μs  ┊ GC (min … max): 0.00% … 94.39%
 Time  (median):     71.279 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   74.554 ns ± 56.606 ns  ┊ GC (mean ± σ):  2.72% ±  3.43%

            ▂▆▇██▇▆▄▂▁▁▁▂▁                    ▁▂▂▁▁           ▂
  ▂▃▃▂▂▂▃▂▂▇███████████████▆▅▅▄▅▅▄▅▂▅▄▆▆▆▄▅▅▆███████▇▆▅▆▅▆▆██ █
  64.3 ns      Histogram: log(frequency) by time      92.8 ns <

 Memory estimate: 64 bytes, allocs estimate: 2.

julia> r = 2.0:8.0
2.0:1.0:8.0

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 994 evaluations.
 Range (min … max):  32.466 ns … 62.311 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     32.909 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   33.192 ns ±  1.534 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▂▆█▄▆                                                     ▁ ▂
  █████▅▅▄▄▆▇▆▄▆▆▇▆▆▆▁▅▅▅▆▇▆▇▅▅▅▄▄▃▃▃▄▃▃▁▃▃▁▁▃▁▃▃▃▃▁▁▃▁▁▄▁▃▅█ █
  32.5 ns      Histogram: log(frequency) by time      42.6 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

If `r` is a range with integers, then the current `master` branch wins, but if `r` is a range with floats, then this PR wins.

>`findall(∈($(1..3)), $(2:8))` being significantly slower than without dollars, for both IntervalSets versions. This is actually the first time I see `$` making benchmark slower. But it occurs with very simple Base code as well: `@btime 1:2:1000` gives 2 ns, `@btime 1:$(2):1000` 10 ns.
Also, very surprising that allocations appear with dollars in benchmark. I don't understand this at all.

I also don't understand the behavior well, but I thought it would be better to use variables such as `i` and `r` to clarify the difference in the performances.
[BenchmarkTools.jl documentation](https://juliaci.github.io/BenchmarkTools.jl/stable/manual/#Interpolating-values-into-benchmark-expressions)

**Comment by hyrodium** [comment]
Maybe, it seems better to split this PR into

* Add `rev` keyword argument in `searchsorted_interval` function.
* Simplify `findall` function.

to merge this PR quickly.

**Comment by hyrodium** [comment]
The `searchsorted` function take O(log(n)) time, but the original `findall` function take O(1) time.

**Bedore this PR**
```julia
julia> using IntervalSets, BenchmarkTools

julia> i = 1..3
1..3

julia> r = 2:8
2:8

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.332 ns … 37.027 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.442 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.741 ns ±  1.556 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▅                  ▁                                     ▁ ▁
  ██▄▄▄▃▃▁▄▅▄▃▄▄▅▄▆▅▃██▅▄▁▅▃▁▁▄▃▁▁▃▄▁▃▃▁▃▁▄▃▁▃▁▁▁▃▁▁▁▁▃▁▁▃▁▆█ █
  11.3 ns      Histogram: log(frequency) by time      21.7 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> r = -2000000:80000000
-2000000:80000000

julia> @benchmark findall(∈($i), $r)  # Benchmark result remains mostly the same.
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  12.475 ns … 22.756 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     12.567 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.619 ns ±  0.280 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▅ ▅▁       ▄ █▂▄  ▂   ▃ ▅            ▆▄       ▁ ▃          ▂
  ▇█▁██▁▅▆▇▁▁▇█▁███▅▁██▁▇█▁█▁▁▁▁▁▁▁▃▁▄█▁██▁▇▇▄▇▁▄█▁█▁▁▃▁▃▃▁▁▄ █
  12.5 ns      Histogram: log(frequency) by time      12.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**After this PR**
```julia
julia> using IntervalSets, BenchmarkTools

julia> i = 1..3
1..3

julia> r = 2:8
2:8

julia> @benchmark findall(∈($i), $r)
BenchmarkTools.Trial: 10000 samples with 973 evaluations.
 Range (min … max):  72.635 ns …  1.769 μs  ┊ GC (min … max): 0.00% … 93.18%
 Time  (median):     74.240 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   77.692 ns ± 58.474 ns  ┊ GC (mean ± σ):  2.70% ±  3.43%

   ▄██▇▇▆▄▃▂▁▁▁▁▁▁                         ▁▁▁▁▁ ▁            ▂
  ▆███████████████▇▆▅▅▅▅▅▅▅▅▄▄▅▅▄▁▁▅▅▃▃▃▄█████████▇▇▇▆▆▅▆▆▇▇█ █
  72.6 ns      Histogram: log(frequency) by time      96.3 ns <

 Memory estimate: 64 bytes, allocs estimate: 2.

julia> r = -2000000:80000000
-2000000:80000000

julia> @benchmark findall(∈($i), $r)  # Takes much time than small r.
BenchmarkTools.Trial: 10000 samples with 967 evaluations.
 Range (min … max):  81.032 ns …  2.237 μs  ┊ GC (min … max): 0.00% … 95.78%
 Time  (median):     84.948 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   90.104 ns ± 88.516 ns  ┊ GC (mean ± σ):  4.38% ±  4.27%

   ▄▇▇█████▇▆▄▃▂▂▁                             ▁ ▁            ▃
  ▆█████████████████▇▆▇▆▆▇▇▆▅▆▃▁▅▅▁▁▃▃▁▃▅▁▅▆▆████████▇▆▇▇▇██▇ █
  81 ns        Histogram: log(frequency) by time       116 ns <

 Memory estimate: 96 bytes, allocs estimate: 4.
```

**Comment by hyrodium** [comment]
>"Quickly" is loooong gone already :)

Sorry for that delay :smile: 

>Not sure if I understand you here... Both searchsorted and searchsorted_interval definitely take O(1) time for ranges:

The searchsorted family functions use binary search, so it increases computing time with the length of `r`.

```julia
julia> r = -2:8
-2:8

julia> @benchmark searchsortedfirst($r,3.3,first($r),last($r),$o)
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  4.919 ns … 14.187 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.959 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.963 ns ±  0.156 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                        █                     
  ▂▁▁▁▁▁▁▁▁▅▄▁▁▁▁▁▁▁▁▅▁▁▁▁▁▁▁▁▂▂▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▇▆▁▁▁▁▁▁▁▁▃ ▂
  4.92 ns        Histogram: frequency by time        4.98 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> r = -20:80
-20:80

julia> @benchmark searchsortedfirst($r,3.3,first($r),last($r),$o)
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  10.721 ns … 27.479 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     10.801 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   10.849 ns ±  0.465 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▆          █                                                
  ▃█▁▇▂▁▂▂▂▂▁▆█▁▃▂▂▂▁▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▁▁▂▂▄▁▃▃▁▁▁▂▂▁▃▃▁▄▂ ▂
  10.7 ns         Histogram: frequency by time        11.1 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> r = -200:800
-200:800

julia> @benchmark searchsortedfirst($r,3.3,first($r),last($r),$o)
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  17.026 ns … 45.697 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     17.348 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   17.412 ns ±  0.465 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄ ▂    █ ▆▆▃▁▁                                              ▂
  █▁██▁▁▅███████▁▃▄▃▄▅▁▁▃▁▁▁▃▃▃▃▃▁▅▁▁▄▁▄▁▄▁▄▅▄▅▁▅▅▃▃▄▃▅▁▁▃▃▄▅ █
  17 ns        Histogram: log(frequency) by time      19.5 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> r = -2000:8000
-2000:8000

julia> @benchmark searchsortedfirst($r,3.3,first($r),last($r),$o)
BenchmarkTools.Trial: 10000 samples with 996 evaluations.
 Range (min … max):  24.282 ns … 48.344 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     24.474 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   24.548 ns ±  0.618 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

     █                                                         
  ▃▁▄█▂▅▂▂▂▂▂▁▁▂▂▂▂▁▂▁▁▂▂▁▁▂▂▁▁▂▂▂▂▂▂▁▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂ ▂
  24.3 ns         Histogram: frequency by time        27.7 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

I'm not sure how the compiler or BenchmarkTools.jl optimizes in `searchsorted_interval`, but ideally, the performance should depend on the length of `r`.

**Comment by hyrodium** [comment]
>`searchsorted` definitely doesn't use binary search for ranges by default

Oh, that was my mistake. Thanks for the clarification.

**Review by hyrodium — COMMENTED** [review]
Sorry for the delayed review and my imprecise comments. I think I now understand all of your points.

This PR fixes the `searchsorted_interval` function for the `AbstractVector` input with `-0.0`, but the change introduces a new bug with `-0.0` in `Inteval`.
```julia
julia> using IntervalSets

julia> searchsorted_interval([-2., -0., 0., 2., 3.], 0.0..3.0)
2:5

julia> searchsorted_interval([-2., -0., 0., 2., 3.], -0.0..3.0)
3:5
```
We also have the following problem with Unitful, so it seems better to use `lt = <` with `searchsorted`-family functions. The performance problem can be avoided with specifying `X::AbstractRange` instead of specifying `x:::AbstractFloat`.

File: src/findall.jl [hunk 1]
```diff
@@ -104,7 +67,22 @@ julia> searchsorted_interval(Float64[], 1..3)
 """
 function searchsorted_interval end
 
-searchsorted_interval(X, i::Interval{:closed, :closed}) = searchsortedfirst(X, leftendpoint(i))     :searchsortedlast(X, rightendpoint(i))
-searchsorted_interval(X, i::Interval{:closed,   :open}) = searchsortedfirst(X, leftendpoint(i))     :(searchsortedfirst(X, rightendpoint(i)) - 1)
-searchsorted_interval(X, i::Interval{  :open, :closed}) = (searchsortedlast(X, leftendpoint(i)) + 1):searchsortedlast(X, rightendpoint(i))
-searchsorted_interval(X, i::Interval{  :open,   :open}) = (searchsortedlast(X, leftendpoint(i)) + 1):(searchsortedfirst(X, rightendpoint(i)) - 1)
+function searchsorted_interval(X, i::Interval{L, R}; rev=false) where {L, R}
```
[the remark this answers is not in the document]
Comment (hyrodium) [reply src/findall.jl:70]: This can be fixed in another PR.

File: src/findall.jl [hunk 2]
```diff
@@ -104,7 +67,22 @@ julia> searchsorted_interval(Float64[], 1..3)
...
+_searchsorted_begin(X, x, ::Val{:open  }; rev) = searchsortedlast( X, x; rev) + 1
+_searchsorted_end(  X, x, ::Val{:closed}; rev) = searchsortedlast( X, x; rev)
+_searchsorted_end(  X, x, ::Val{:open  }; rev) = searchsortedfirst(X, x; rev) - 1
+
+# Base searchsorted functions use the `isless` comparator, while interval membership uses `<`.
+# They differ for floating point zeros. Here, we change the zero sign when needed, so that the searchsorted result agrees with `<`.
+_searchsorted_begin(X, x::AbstractFloat, ::Val{:closed}; rev) = searchsortedfirst(X, iszero(x) && (x < zero(x)) == rev ? -x : x; rev)
+_searchsorted_begin(X, x::AbstractFloat, ::Val{:open  }; rev) = searchsortedlast( X, iszero(x) && (x < zero(x)) != rev ? -x : x; rev) + 1
+_searchsorted_end(  X, x::AbstractFloat, ::Val{:closed}; rev) = searchsortedlast( X, iszero(x) && (x < zero(x)) != rev ? -x : x; rev)
+_searchsorted_end(  X, x::AbstractFloat, ::Val{:open  }; rev) = searchsortedfirst(X, iszero(x) && (x < zero(x)) == rev ? -x : x; rev) - 1
```
[the remark this answers is not in the document]
Comment (hyrodium) [reply src/findall.jl:88]: The above issue already exists on the current master, so can be fixed in another PR.

**Review by hyrodium — APPROVED** [review]
@aplavin Sorry for the late. I will merge this PR in a week.

**Comment by hyrodium** [comment]
>To avoid the performance regression for ranges.

I personally feel we can ignore performance degressions in old Julia versions (if the degression is not that much), so it would be okay to merge this PR without `if VERSION < ...` not to increase the maintenance cost.

>I don't really feel like crafting separate tests for all edge cases of `searchsorted_interval(rev=false/true)` - `findall` tests cover them already, and using `searchsorted_interval` in `findall` only makes sense.

No problem with this direction!

I will merge this PR after https://github.com/JuliaLang/julia/pull/50365 is merged.

**Comment by hyrodium** [comment]
The PR https://github.com/JuliaLang/julia/pull/50365 was merged, so I have merged this PR.

@aplavin 
Thank you for the PR and for your patience while waiting for the review & merge!
